### PR TITLE
Pinot Connector: Adding limit to push down order by for broker query

### PR DESCRIPTION
**Pull request by fx19880617**
```
== RELEASE NOTES ==

Pinot Changes
* Adding config: `pinot.pushdown-topn-broker-queries` to support pushing down TOPN queries.

Presto doesn't retain the order of query response from Pinot for large number of records returned. So only enable this when doing small group by and order by results
```

### CODE REVIEW

**Comment by fx19880617** [comment]
@agrawaldevesh this is the issue I mentioned about pushing down order by.

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGeneratorContext.java [hunk 1]
```diff
@@ -255,7 +255,6 @@ public PinotQueryGeneratorContext withTopN(LinkedHashMap<VariableReferenceExpres
             checkSupported(!hasAggregation(), "Pinot doesn't support ordering on top of the aggregated data");
         }
         int intLimit = checkForValidLimit(limit);
-
```
Comment (agrawaldevesh) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGeneratorContext.java:258]: unrelated

File: presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/query/TestPinotQueryGeneratorSql.java [hunk 2]
```diff
@@ -158,6 +160,36 @@ public void testAggregationWithOrderByPushDownInTopN()
...
+        TopNNode topN = new TopNNode(
+                planBuilder.getIdAllocator().getNextId(),
+                agg,
+                limitToPushdownOrderByForBrokerQueries,
+                new OrderingScheme(ImmutableList.of(new Ordering(variable("sum_fare"), SortOrder.ASC_NULLS_FIRST))),
+                TopNNode.Step.SINGLE);
+        testPinotQuery(
+                pinotConfig,
+                topN,
+                "SELECT city, sum(fare) FROM realtimeOnly GROUP BY city ORDER BY sum(fare) LIMIT " + limitToPushdownOrderByForBrokerQueries,
```
Comment (agrawaldevesh) [inline presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/query/TestPinotQueryGeneratorSql.java:184]: Can you also test a non limited order by ? Which should perhaps be a SortNode ? Does that also have the same problem as a TopNNode ?

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java [hunk 3]
```diff
@@ -468,4 +470,18 @@ public PinotConfig setUsePinotSqlForBrokerQueries(boolean usePinotSqlForBrokerQu
         this.usePinotSqlForBrokerQueries = usePinotSqlForBrokerQueries;
         return this;
     }
+
+    public int getLimitToPushdownOrderByForBrokerQueries()
+    {
+        return this.limitToPushdownOrderByForBrokerQueries;
```
Comment (agrawaldevesh) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:476]: nit: limitThreshholdForTopNBrokerQueries ? 

To be more explicit that this covers order-by-limit and not unlimited order by. 

(And corresponding changes to the other DEFAULT_ and derived strings from this config name).

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java [hunk 4]
```diff
@@ -38,6 +38,7 @@
     public static final int DEFAULT_MIN_CONNECTIONS_PER_SERVER = 10;
     public static final int DEFAULT_THREAD_POOL_SIZE = 30;
     public static final int DEFAULT_NON_AGGREGATE_LIMIT_FOR_BROKER_QUERIES = 25_000;
+    public static final int DEFAULT_LIMIT_TO_PUSHDOWN_ORDER_BY_FOR_BROKER_QUERIES = 5_000;
```
Comment (agrawaldevesh) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: Would love to know more about where this 5K was inferred from ? Is there a Presto config we can refer to instead here ? 

Like do we know where and according to what config is Presto violating this order ? If it is a simple threshhold, perhaps we should refer to that threshhold here instead of creating our own, to avoid them being out of sync.

Comment (fx19880617) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: This is from my local test, I observed the block to be 7kish.

Comment (agrawaldevesh) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: Basically, this seems a bit brittle. I am wondering if we can be more surer in when to trigger this. Like should this be disabled if the connector would return more than a page ? (Then perhaps we should fail the query if order-by-limit pushdown was done but more than one page is being returned because of a wide result set). 

Or is there a config ? 

I am basically wary of the case when the limit is lower than this number, we push down order-by-limit and still see the issue you faced originally or incorrect ordering. 

Disabling this outright would also be okay with me because that is also deterministic.

Comment (fx19880617) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: I see, then in this case I can just change it to 1 to disable it.
I'm not sure if Presto has config to define the page size.

File: presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/query/TestPinotQueryGeneratorSql.java (hunk 2 above)
Comment (fx19880617) [reply presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/query/TestPinotQueryGeneratorSql.java:184]: Non limit orderby query will generate a SortNode, which is not pushdown-able, so it has no problem like this.

Comment (fx19880617) [reply presto-pinot-toolkit/src/test/java/com/facebook/presto/pinot/query/TestPinotQueryGeneratorSql.java:184]: Also there is a default limit applied to Pinot query/

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java (hunk 3 above)
Comment (fx19880617) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java]: done.

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSessionProperties.java [hunk 5]
```diff
@@ -40,6 +40,7 @@
     private static final String USE_DATE_TRUNC = "use_date_trunc";
     private static final String USE_PINOT_SQL_FOR_BROKER_QUERIES = "use_pinot_sql_for_broker_queries";
     private static final String NON_AGGREGATE_LIMIT_FOR_BROKER_QUERIES = "non_aggregate_limit_for_broker_queries";
+    private static final String LIMIT_TO_PUSHDOWN_ORDER_BY_FOR_BROKER_QUERIES = "limit_to_pushdown_order_by_for_broker_queries";
```
Comment (agrawaldevesh) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSessionProperties.java:43]: Reword this one too please

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGenerator.java [hunk 6]
```diff
@@ -245,12 +245,14 @@ public String toString()
...
         private final boolean forbidBrokerQueries;
         private final boolean useSqlSyntax;
+        private final int limitToPushdownOrderByForBrokerQueries;
 
         protected PinotQueryPlanVisitor(ConnectorSession session)
         {
             this.session = session;
             this.forbidBrokerQueries = PinotSessionProperties.isForbidBrokerQueries(session);
             this.useSqlSyntax = PinotSessionProperties.isUsePinotSqlForBrokerQueries(session);
+            this.limitToPushdownOrderByForBrokerQueries = PinotSessionProperties.getLimitToPushdownOrderByForBrokerQueries(session);
```
Comment (agrawaldevesh) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGenerator.java:255]: Reword this too please

**Review by agrawaldevesh — COMMENTED** [review]
So effectively we are not pushing down topn node now right ? 

Bummer I wish we can get to the bottom of it, but this fix also suffices.

Eventually if we cannot fix this properly, we should just take out the topn pushdown logic.

**Comment by fx19880617** [comment]
> So effectively we are not pushing down topn node now right ?
> 
> Bummer I wish we can get to the bottom of it, but this fix also suffices.
> 
> Eventually if we cannot fix this properly, we should just take out the topn pushdown logic.

Agreed, this fix will rollback the order by push down.
Once we figure out the fix, we can remove this.

**Review by agrawaldevesh — APPROVED** [review]
LGTM

**Review by agrawaldevesh — APPROVED** [review]
LGTM

**Comment by fx19880617** [comment]
@zhenxiao @highker Could you help merge this ? Thanks!

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java [hunk 7]
```diff
@@ -475,4 +477,18 @@ public PinotConfig setUsePinotSqlForBrokerQueries(boolean usePinotSqlForBrokerQu
         this.usePinotSqlForBrokerQueries = usePinotSqlForBrokerQueries;
         return this;
     }
+
+    public int getLimitThresholdForTopNBrokerQueries()
+    {
+        return this.limitThresholdForTopNBrokerQueries;
```
Comment (zhenxiao) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:483]: return limitThresholdForTopNBrokerQueries;

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java [hunk 8]
```diff
@@ -475,4 +477,18 @@ public PinotConfig setUsePinotSqlForBrokerQueries(boolean usePinotSqlForBrokerQu
...
+    }
+
+    // This is used to not push down Pinot broker queries with ORDER BY clause and large LIMIT.
```
Comment (zhenxiao) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:486]: to not push down -> to limit push down

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java [hunk 9]
```diff
@@ -38,6 +38,7 @@
     public static final int DEFAULT_MIN_CONNECTIONS_PER_SERVER = 10;
     public static final int DEFAULT_THREAD_POOL_SIZE = 30;
     public static final int DEFAULT_NON_AGGREGATE_LIMIT_FOR_BROKER_QUERIES = 25_000;
+    public static final int DEFAULT_LIMIT_THRESHOLD_FOR_TOP_BROKER_QUERIES = 1;
```
Comment (zhenxiao) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: shall we use a bigger number as default? e.g. 1000

**Review by zhenxiao — APPROVED** [review]
thank you, @fx19880617 
LGTM, just a few questions

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java (hunk 9 above)
Comment (fx19880617) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: we had a previous discussion, this value could be varies based on user's query, so it would be better to disable it by default and enable it based on the query.

Comment (agrawaldevesh) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: Let me further clarify: Basically, Its not 100% safe to remove sorting from Presto and delegate that entirely to the connector. We are not sure of the root reason behind this. 

What happen is that on getting these many records (ie limitThreshholdForTopBrokerQueries), Presto might end up reordering them. And the absence of the sort makes this reordering wrong: There is no final sort at the end to correct the required order of the data.

What's more that this "limit" under which the presto might end up reordering can depend on the width of the data returned. So we can't be sure that it should be say 1000. The only thing safe here is to disable sort pushdown entirely . The way that is done here is to only push down if there is already a LIMIT 1 (making the sorting order moot).

Another way to implement the PR would have been to simply remove the TopN pushdown logic entirely and/or boolean config disable it.

Comment (zhenxiao) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: oh, get it. thank you for the explanation, @agrawaldevesh 
if I understand correctly, setting DEFAULT_LIMIT_THRESHOLD_FOR_TOP_BROKER_QUERIES > 1, might trigger incorrect result due to topN pushdown? please fix me if I mis-understand
I am inclined to have a config to disable topN pushdown, instead of relying on default value to guarantee correct behavior. what do you think @fx19880617 @agrawaldevesh

Comment (fx19880617) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: I feel it's better to have the config, rather than disable it.
E.g. I want to query for `select sum(a) as sum_a from myTable order by sum_a desc limit 10`
then it's actually we can push down the entire query to pinot.

Comment (agrawaldevesh) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: One discussion we kept having when we were building this was: 

If the number of rows being returned by Pinot is small, then the sort in Presto has a very low cost and thus there is not much benefit in pushing down the sort. The benefit is only pronounced when we have a potentially high cardinality group by (not a global aggregation like above which produces only a single row), because its only there that we avoid fishing out the whole group-by from Pinot only to sort and select the top-K rows from it in Presto.

But unfortunately, we don't know what value of "K" is safe to be pushed down ? Is it 10, 100, 1000 ? How would our answer change with large widths etc. I wish we really understood the root cause here :-)

Comment (zhenxiao) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: yep, am inclined to have a config to control whether we do topN pushdown, default to be false.
For @fx19880617 's case, we could enable topN pushdown for the safe queries.
what do you think?

Comment (fx19880617) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: right, I feel we should still keep the possibility for users to optimize the query if needed and possible

Comment (fx19880617) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: ah, what you mean is to make this a session config and let user to decide turn it on/off instead of using a limit number ?

Comment (zhenxiao) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java:41]: yes, instead of using a configurable limit number, let's have a boolean session property, which enables topN pushdown, defaults to be false.
so that default will guarantee correctness. For advanced users who know their queries could benefit from pushdown, they could turn it on query by query.
exposing the limit number requires more domain knowledge for users, in addition to the risk of correctness

Comment (fx19880617) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotConfig.java]: done.

**Review by agrawaldevesh — APPROVED** [review]
I glanced over the tests but could not ascertain whether there is a negative test: That broker topn pushdown does not happen if the config is disabled. Can you please ensure that is there.

The code is nice and surgical and should totally work.

**Comment by fx19880617** [comment]
> I glanced over the tests but could not ascertain whether there is a negative test: That broker topn pushdown does not happen if the config is disabled. Can you please ensure that is there.
> 
> The code is nice and surgical and should totally work.

Please see `testDefaultNoTopNByPushdown`. So here we expect the exception to be thrown.

**Review by agrawaldevesh — APPROVED** [review]
Duh ! Thanks for double checking that.

File: presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGenerator.java [hunk 10]
```diff
@@ -515,6 +517,10 @@ public PinotQueryGeneratorContext visitTopN(TopNNode node, PinotQueryGeneratorCo
             requireNonNull(context, "context is null");
             checkSupported(!forbidBrokerQueries, "Cannot push topn in segment mode");
             checkSupported(node.getStep().equals(TopNNode.Step.SINGLE), "Can only push single logical topn in");
+
+            if (!pushdownTopnBrokerQueries) {
```
Comment (zhenxiao) [inline presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGenerator.java:521]: shall we throw exception here?
am expecting:

if (pushdownTopnBrokerQueries) {
    return context.withTopN(...)
}
return context;

Comment (agrawaldevesh) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGenerator.java:521]: Throwing an exception is indeed the right thing to do here: The whole Pinot pushdown framework uses exceptions to know "when to push no further". The exception will be propagated all the way to PinotPushdownPlanOptimizer and will stall future pushdown. The exception trick is used across the entire file and so I would vote for consistency with the rest of the file.

Comment (zhenxiao) [reply presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGenerator.java:521]: get it.
